tests/integration/test_charm.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+
+"""Integration tests for Namespace Node Affinity Operator/Charm."""
+
+import logging
+from pathlib import Path
+from time import sleep
+
+import pytest
+import yaml
+from lightkube import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.models.core_v1 import Container, ContainerPort, PodSpec
+from lightkube.models.meta_v1 import ObjectMeta
+from lightkube.resources.core_v1 import Namespace, Pod
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = "namespace-node-affinity"
+
+
+SETTINGS_YAML_TEMPLATE = """
+{namespace}: |
+    nodeSelectorTerms:
+      - matchExpressions:
+        - key: the-testing-key
+          operator: In
+          values:
+          - the-testing-val1
+      - matchExpressions:
+        - key: the-testing-key2
+          operator: In
+          values:
+          - the-testing-val2
+        """
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build and deploy the charm, asserting on the unit status."""
+    charm_under_test = await ops_test.build_charm(".")
+
+    await ops_test.model.deploy(charm_under_test, application_name=APP_NAME, trust=True)
+
+    # NOTE: idle_period is used to ensure all resources are deployed
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=True, timeout=60 * 10
+    )
+    assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+
+
+LABEL_KEY = "app"
+LABEL_VALUE = "tests-for-namespace-node-affinity"
+TEST_LABEL = {LABEL_KEY: LABEL_VALUE}
+
+
+def create_test_pod_resource(name: str) -> Pod:
+    """Return a sample Pod resource with the given name."""
+    pod = Pod(
+        metadata=ObjectMeta(
+            name=name,
+            labels=TEST_LABEL,
+            # namespace=namespace
+        ),
+        spec=PodSpec(
+            containers=[
+                Container(
+                    name="nginx", image="nginx:1.14.2", ports=[ContainerPort(containerPort=80)]
+                )
+            ]
+        ),
+    )
+    return pod
+
+
+@pytest.fixture()
+def temp_pod_deleter():
+    """Fixture to delete all pods with the test label after a test has run.
+
+    Use this fixture to to ensure no leftover pods remain after a test
+    """
+    # Do nothing on call
+    yield
```

Comment:
Will this prevent the rest of method from running? Why not rid of all other code?

Reply:
The way this works is that when pytest uses the fixture, you'll get everything up to the `yield`.  Then after whatever used the fixture is done, you'll get back to here and complete the rest of the code.  So something like:

```python
def test_something(my_fixture):
  print("I'm in the test")

def my_fixture():
  print("Fixture startup")
  yield
  print("Fixture teardown")
```
would print like:
```
Fixture startup
I'm in the test
Fixture teardown
```

The reason I'm using it here, is I want something that is always guaranteed to execute after the test to clean up any leftover pods.  The teardown code could be in the test itself, but then if we hit an exception in the middle of the test the teardown would never run.  This way, the fixture essentially works as a `finally` block on the test

Reply:
Great!